pkg/mcs/resource_manager/server/token_bukets.go

```diff
@@ -38,107 +44,227 @@ type GroupTokenBucket struct {
 	Settings              *rmpb.TokenLimitSettings `json:"settings,omitempty"`
 	GroupTokenBucketState `json:"state,omitempty"`
 }
 
+func (gtb *GroupTokenBucket) setState(state *GroupTokenBucketState) {
+	gtb.Tokens = state.Tokens
+	gtb.LastUpdate = state.LastUpdate
+	gtb.Initialized = state.Initialized
+	log.Info("update group token bucket state", zap.Any("state", state))
+}
+
+// TokenSlot is used to split a token bucket into multiple slots to
+// server different clients within the same resource group.
+type TokenSlot struct {
+	// settings is the token limit settings for the slot.
+	settings *rmpb.TokenLimitSettings
+	// requireTokensSum is the number of tokens required.
+	requireTokensSum float64
+	// tokenCapacity is the number of tokens in the slot.
+	tokenCapacity     float64
+	lastTokenCapacity float64
+
+	lastUpdate *time.Time
+}
+
 // GroupTokenBucketState is the running state of TokenBucket.
 type GroupTokenBucketState struct {
-	Tokens      float64    `json:"tokens,omitempty"`
+	Tokens float64 `json:"tokens,omitempty"`
+	// ClientUniqueID -> TokenSlot
+	tokenSlots                 map[uint64]*TokenSlot
+	clientConsumptionTokensSum float64
+
 	LastUpdate  *time.Time `json:"last_update,omitempty"`
 	Initialized bool       `json:"initialized"`
 	// settingChanged is used to avoid that the number of tokens returned is jitter because of changing fill rate.
 	settingChanged bool
 }
 
 // Clone returns the copy of GroupTokenBucketState
-func (s *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+func (gts *GroupTokenBucketState) Clone() *GroupTokenBucketState {
+	tokenSlots := make(map[uint64]*TokenSlot)
+	for id, tokens := range gts.tokenSlots {
+		tokenSlots[id] = tokens
+	}
+	var lastUpdate *time.Time
+	if gts.LastUpdate != nil {
+		newLastUpdate := *gts.LastUpdate
+		lastUpdate = &newLastUpdate
+	}
 	return &GroupTokenBucketState{
-		Tokens:      s.Tokens,
-		LastUpdate:  s.LastUpdate,
-		Initialized: s.Initialized,
+		Tokens:                     gts.Tokens,
+		LastUpdate:                 lastUpdate,
+		Initialized:                gts.Initialized,
+		tokenSlots:                 tokenSlots,
+		clientConsumptionTokensSum: gts.clientConsumptionTokensSum,
+	}
+}
+
+func (gts *GroupTokenBucketState) cleanupAssignTokenSum() {
+	gts.clientConsumptionTokensSum = 0
+	for _, slot := range gts.tokenSlots {
+		slot.requireTokensSum = 0
+	}
+}
+
+func (gts *GroupTokenBucketState) balanceSlotTokens(
+	clientUniqueID uint64,
+	settings *rmpb.TokenLimitSettings,
+	consumptionToken float64) {
+	now := time.Now()
+	slot, exist := gts.tokenSlots[clientUniqueID]
+	if !exist {
+		gts.cleanupAssignTokenSum()
+		slot = &TokenSlot{}
+		gts.tokenSlots[clientUniqueID] = slot
+	} else if gts.clientConsumptionTokensSum >= maxAssignTokens {
+		gts.cleanupAssignTokenSum()
+	}
+	slot.lastUpdate = &now
+	slot.requireTokensSum += consumptionToken
+	gts.clientConsumptionTokensSum += consumptionToken
+
+	evenRatio := 1 / float64(len(gts.tokenSlots))
```

Comment:
should make sure `len(gts.tokenSlots) > 0`

Reply:
Before starting the code here, tokenSlots will check request id first, 
if it does not include client id, it will add it, which means not empty now.
Is the process of adding an error?